src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignerInfoTests.netcoreapp.cs

```diff
@@ -77,6 +77,55 @@ public static void SignerInfo_RemoveUnsignedAttribute_RemoveCounterSignature()
             Assert.Equal(secondSignerCounterSignature, cms.SignerInfos[0].CounterSignerInfos[0].GetSignature());
         }
 
+        [Fact]
+        public static void SignerInfo_RemoveUnsignedAttributeFromCounterSigner_Removes()
+        {
+            SignedCms cms = new SignedCms();
+            cms.Decode(SignedDocuments.OneRsaSignerTwoRsaCounterSigners);
+
+            AsnEncodedData attribute1 = CreateTimestampToken(1);
+            AsnEncodedData attribute2 = CreateTimestampToken(2);
+
+            SignerInfo counterSigner = cms.SignerInfos[0].CounterSignerInfos[0];
+            counterSigner.AddUnsignedAttribute(attribute1);
+            counterSigner.AddUnsignedAttribute(attribute2);
+
+            ReReadSignedCms(ref cms);
+
+            SignerInfo signer = cms.SignerInfos[0];
+
+            // attributes got sorted, therefore number changed
+            Assert.Equal(2, signer.CounterSignerInfos.Count);
+            Assert.Equal(1, signer.CounterSignerInfos[1].UnsignedAttributes.Count);
+            Assert.Equal(2, signer.CounterSignerInfos[1].UnsignedAttributes[0].Values.Count);
+
+            counterSigner = signer.CounterSignerInfos[1];
+            counterSigner.RemoveUnsignedAttribute(counterSigner.UnsignedAttributes[0].Values[0]);
+
+            // We didn't modify existing instances
+            Assert.Equal(2, signer.CounterSignerInfos.Count);
+            Assert.Equal(1, counterSigner.UnsignedAttributes.Count);
+            Assert.Equal(2, counterSigner.UnsignedAttributes[0].Values.Count);
+
+            // We did modify the state
+            Assert.Equal(1, cms.SignerInfos[0].CounterSignerInfos[1].UnsignedAttributes.Count);
+            Assert.Equal(1, cms.SignerInfos[0].CounterSignerInfos[1].UnsignedAttributes[0].Values.Count);
```

Comment:
what about signer.CounterSignerInfos[1].UnsignedAttributes[0].Count?

Reply:
this one is weird. the parent gets updated after first add but after second add since it's not parent anymore it doesn't get updated (same for removing) - this feels super weird but as we discussed we're not gonna change this for 3.1 so leaving as is.